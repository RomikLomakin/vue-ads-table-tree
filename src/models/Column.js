export default class Column {
    constructor (properties) {
        Object.assign(this, properties);
    }

    set property (property) {
        this._property = property;
    }

    get property () {
        if (!this.hasOwnProperty('_property')) {
            this._property = '';
        }

        return this._property;
    }

    set order (order) {
        this._order = order;
    }

    get order () {
        if (!this.hasOwnProperty('_order')) {
            return 0;
        }

        return this._order;
    }

    set collapseIcon (collapseIcon) {
        this._collapseIcon = collapseIcon;
    }

    get collapseIcon () {
        if (!this.hasOwnProperty('_collapseIcon')) {
            this._collapseIcon = false;
        }

        return this._collapseIcon;
    }

    set sortable (sortable) {
        this._sortable = sortable;
    }

    get sortable () {
        if (!this.hasOwnProperty('_sortable')) {
            this._sortable = false;
        }

        return this._sortable;
    }

    set direction (direction) {
        if (![true, false, null].includes(direction)) {
            throw new Error(
                'Sort direction has to be true (asc), false (desc) or null (no sorting)',
            );
        }

        this._direction = direction;
    }

    get direction () {
        if (!this.hasOwnProperty('_direction')) {
            this._direction = null;
        }

        return this._direction;
    }

    set filterable (filterable) {
        this._filterable = filterable;
    }

    get filterable () {
        if (!this.hasOwnProperty('_filterable')) {
            this._filterable = false;
        }

        return this._filterable;
    }

    sort (order) {
        this.direction = !this.direction;
        this.order = order;
    }
}
