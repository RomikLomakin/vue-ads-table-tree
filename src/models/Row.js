import RowCollection from '../collections/RowCollection';

export default class Row {
    constructor (properties) {
        this.children = new RowCollection();
        this.processedChildren = null;
        this.showChildren = false;
        this.hasChildren = false;
        this.childrenLoading = false;
        Object.assign(this, properties);
    }

    set parent (parent) {
        if (!(parent instanceof Row)) {
            throw new Error(
                'Parent of a row has to be a row'
            );
        }

        this._parent = parent;
    }

    get parent () {
        return this._parent;
    }

    set children (children) {
        this._children = children instanceof RowCollection ? children : new RowCollection(children);
        this._children.items
            .forEach(childRow => {
                childRow.parent = this;
            });

        this.hasChildren = !this._children.isEmpty();
    }

    get children () {
        return this._children;
    }

    set childrenLoading (childrenLoading) {
        this._childrenLoading = childrenLoading;
    }

    get childrenLoading () {
        return this._childrenLoading;
    }

    set processedChildren (processedChildren) {
        this._processedChildren = processedChildren;

        // wss voor async calls.
        // this._processedChildren.items
        //     .forEach(filteredChild => {
        //         filteredChild.parent = this;
        //     });
    }

    get processedChildren () {
        return this._processedChildren === null ? this.children : this._processedChildren;
    }

    set hasChildren (hasChildren) {
        this._hasChildren = hasChildren;
    }

    get hasChildren () {
        return this._hasChildren;
    }

    set showChildren (showChildren) {
        this._showChildren = showChildren;
    }

    get showChildren () {
        return this._showChildren;
    }

    get properties () {
        return Object.getOwnPropertyNames(this)
            .filter(property => {
                return property[0] !== '_';
            });
    }

    toggleChildren () {
        this.showChildren = !this.showChildren;
    }

    childrenLoaded () {
        return this.hasChildren && !this.children.isEmpty();
    }

    loadChildren () {
        return this.showChildren && this.hasChildren && this.children.isEmpty();
    }

    countParents () {
        if (this.parent) {
            return this.parent.countParents() + 1;
        }

        return 0;
    }
}
