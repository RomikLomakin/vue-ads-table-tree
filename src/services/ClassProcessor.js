
export default class ClassProcessor {
    constructor (classes, totalColumns) {
        this.totalColumns = totalColumns;
        this.classes = classes;
    }

    set classes (classes) {
        this._classes = classes;
        this.processClasses();
    }

    get classes () {
        return this._classes || [];
    }

    set totalRows (totalRows) {
        this._totalRows = totalRows;
        this.processClasses();
    }

    get totalRows () {
        return this._totalRows || 0;
    }

    set totalColumns (totalColumns) {
        this._totalColumns = totalColumns;
        this.processClasses();
    }

    get totalColumns () {
        return this._totalColumns || 0;
    }

    get processedClasses () {
        return this._processedClasses;
    }

    processClasses () {
        this._processedClasses = Object.keys(this.classes)
            .filter(key => key.includes('/'))
            .map(key => {
                let type = key.split('/');
                return {
                    rows: this.toRange(type[0], this.totalRows),
                    columns: this.toRange(type[1], this.totalColumns),
                    value: this.classes[key],
                };
            });
    }

    toRange (selector, total) {
        if (selector === '' || total === 0) {
            return [];
        }

        switch (selector) {
        case 'all':
            return Array.from(Array(total).keys());
        case 'even':
            return Array.from(Array(total).keys()).filter(item => (item % 2) === 0);
        case 'odd':
            return Array.from(Array(total).keys()).filter(item => (++item % 2) === 0);
        }

        return [].concat(
            ...selector.split(',')
                .map(selector => selector.trim())
                .map(selector => {
                    if (selector.includes('_')) {
                        let range = selector.split('_')
                            .map(index => Number.parseInt(index))
                            .map(index => index < 0 ? total + index : index);

                        if (range[0] < 0 || range[1] < 0 || range[0] > range[1]) {
                            return null;
                        }

                        return Array.from(Array(range[1] - range[0]).keys())
                            .map(number => number + range[0]);
                    }

                    return Number.parseInt(selector);
                })
                .filter(selector => selector !== null)
        );
    }

    process (rowIndex = null, columnIndex = null, ...args) {
        return this.processedClasses
            .map(classes => {
                if (
                    (rowIndex === null && columnIndex === null) ||
                    (columnIndex === null && classes.columns.length > 0) ||
                    (rowIndex === null && classes.rows.length > 0) ||
                    (columnIndex !== null && !classes.columns.includes(columnIndex)) ||
                    (rowIndex !== null && !classes.rows.includes(rowIndex))
                ) {
                    return null;
                }

                return ClassProcessor.processValue(classes.value, ...args);
            })
            .filter(classes => classes)
            .reduce((result, classes) => Object.assign(result, classes), {});
    }

    static processValue (classes, ...args) {
        if (classes instanceof Function) {
            return classes(...args);
        }

        if (classes) {
            return classes;
        }

        return {};
    }

    processFixed (classes, columnIndex, ...args) {
        if (!classes) {
            return {};
        }

        return Object.keys(classes)
            .map(key => {
                if (!this.toRange(key, this.totalColumns).includes(columnIndex)) {
                    return null;
                }

                return ClassProcessor.processValue(classes[key], ...args);
            })
            .filter(classes => classes)
            .reduce((result, classes) => Object.assign(result, classes), {});
    }
}
