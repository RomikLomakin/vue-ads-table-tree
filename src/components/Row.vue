<template>
    <tr
        :class="rowClasses"
    >
        <vue-ads-cell
            v-for="(column, key) in columns.items"
            slot="slot(column)"
            :key="key"
            :row-index="$vnode.key"
            :row="row"
            :column="column"
            :classes="classes"
            @toggleChildren="$emit('toggleChildren');"
        />
    </tr>
</template>

<script>
import Row from '../models/Row';
import VueAdsCell from './Cell';

import ClassProcessor from '../services/ClassProcessor';

export default {
    name: 'VueAdsRow',

    components: {
        VueAdsCell,
    },

    props: {
        row: {
            type: Row,
            required: true,
        },

        columns: {
            type: Object,
            required: true,
        },

        slots: {
            type: Object,
            required: false,
            default: () => { return {}; },
        },

        classes: {
            type: ClassProcessor,
            required: true,
        },
    },

    computed: {
        rowClasses () {
            return Object.assign(
                this.classes.process(this.$vnode.key + 1, null, this.row),
                this.row.classes ? ClassProcessor.processValue(this.row.classes.row, this.row) : {}
            );
        },
    },

    methods: {
        columnSlot (column) {
            return this.slots[column.property + '_' + this.row[column.property]] ||
                this.slots[column.property] ||
                null;
        },
    },
};
</script>
