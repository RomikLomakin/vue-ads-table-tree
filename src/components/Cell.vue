<script>
import Row from '../models/Row';
import Column from '../models/Column';

import VueAdsChildrenButton from './ChildrenButton';

import ClassProcessor from '../services/ClassProcessor';

export default {
    name: 'VueAdsCell',

    components: {
        VueAdsChildrenButton,
    },

    props: {
        row: {
            type: Row,
            required: true,
        },

        rowIndex: {
            type: Number,
            required: true,
        },

        column: {
            type: Column,
            required: true,
        },

        classes: {
            type: ClassProcessor,
            required: true,
        },

        columnSlot: {
            type: Function,
            required: false,
            default: null,
        },
    },

    computed: {
        cellClasses () {
            return Object.assign(
                {
                    'vue-ads-px-4': true,
                    'vue-ads-py-2': true,
                    'vue-ads-text-sm': true,
                },
                this.classes.process(null, this.key, this.column),
                this.classes.process(this.rowIndex + 1, this.key, this.row, this.column),
                this.classes.processFixed(this.row.classes, this.key, this.row, this.column)
            );
        },

        style () {
            return {
                'padding-left': (1 + this.column.collapseIcon * this.row.countParents() * 1.5) + 'rem',
            };
        },

        key () {
            return this.$vnode.key || 0;
        },
    },

    methods: {
        value (createElement) {
            let elements = [];

            if (this.column.collapseIcon && this.row.hasChildren) {
                elements.push(createElement(VueAdsChildrenButton, {
                    props: {
                        expanded: this.row.showChildren,
                        loading: this.row.loading,
                    },
                    nativeOn: {
                        click: this.toggleChildren,
                    },
                }),);
            }

            if (this.columnSlot) {
                elements.push(this.columnSlot({
                    row: this.row.properties,
                }));
            } else if (this.column.property && this.row.properties.hasOwnProperty(this.column.property)) {
                elements.push(this.row.properties[this.column.property]);
            }

            return elements.length > 0 ? elements : [
                '',
            ];
        },

        toggleChildren () {
            this.$emit('toggleChildren');
        },
    },

    render (createElement) {
        return createElement('td', {
            class: this.cellClasses,
            style: this.style,
        }, [
            createElement('span', {}, this.value(createElement)),
        ]);
    },
};
</script>
