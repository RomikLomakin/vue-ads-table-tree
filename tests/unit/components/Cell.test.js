import { shallowMount } from '@vue/test-utils';

import Cell from '../../../src/components/Cell';
import Row from '../../../src/models/Row';
import Column from '../../../src/models/Column';
import ClassProcessor from '../../../src/services/ClassProcessor';

describe('Cell', () => {
    let cell;

    beforeEach(() => {
        cell = shallowMount(Cell, {
            propsData: {
                row: new Row({
                    firstName: 'arne',
                }),
                column: new Column({
                    property: 'firstName',

                }),
                rowIndex: 0,
                classes: new ClassProcessor({}, 0),
            },
        });
    });

    it('returns the default cell classes', () => {
        expect(cell.vm.cellClasses).toEqual({
            'vue-ads-px-4': true,
            'vue-ads-py-2': true,
            'vue-ads-text-sm': true,
        });

        expect(cell.vm.style['padding-left']).toBe('1rem');

        expect(cell.html()).toMatchSnapshot();
    });

    it('changes the padding if the number of parents changes', () => {
        cell.setProps({
            row: new Row({
                parent: new Row(),
            }),

            column: new Column({
                collapseIcon: true,
            }),
        });

        expect(cell.vm.style['padding-left']).toBe('2.5rem');
    });

    it('adds a toggle children button if the row has children', () => {
        cell.setProps({
            row: new Row({
                hasChildren: true,
            }),
        });

        expect(cell.html()).toMatchSnapshot();
    });

    it('is empty with no matching properties', () => {
        cell.setProps({
            row: new Row({
                lastName: 'de smedt',
            }),
        });

        expect(cell.text()).toBe('');
    });

    it('emits toggle children, when calling toggle children', () => {
        cell.vm.toggleChildren();

        expect(cell.emitted().toggleChildren).toBeTruthy();
    });
});
