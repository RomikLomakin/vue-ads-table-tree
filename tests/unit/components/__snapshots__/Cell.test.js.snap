// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell adds a toggle children button if the row has children 1`] = `<td class="vue-ads-px-4 vue-ads-py-2 vue-ads-text-sm vue-ads-w-auto" style="padding-left: 1rem;"><span><i class="fa vue-ads-pr-2 vue-ads-cursor-pointer fa-plus-square"></i></span></td>`;

exports[`Cell returns the default cell classes 1`] = `<td class="vue-ads-px-4 vue-ads-py-2 vue-ads-text-sm vue-ads-w-auto" style="padding-left: 1rem;"><span>arne</span></td>`;
